File: mergesort.py
import math

def main():
    arr = [20, 60, 30, 70, 40, 50, 10]
    print(arr)
    # left_index = 0
    # right_index = len(arr) - 1
    # mergeSort(arr, left_index, right_index)
    mergeSort(arr)
    print(arr)

# def mergeSort (arr, left_index, right_index):
#     #base case
#     if left_index == right_index:
#         return
    
#     middle_index = int(math.floor((left_index + right_index) / 2))
    
#     #recursive case
#     mergeSort(arr, left_index, middle_index)
#     mergeSort(arr, middle_index + 1, right_index)

#     #create temporary arrays for left and right half
#     left_half_size = middle_index - left_index + 1
#     temp_left_array = []
#     for i in range(left_half_size):
#         temp_left_array.append(arr[left_index + i])
#     right_half_size = right_index - middle_index
#     temp_right_array = []
#     for i in range(right_half_size):
#         temp_right_array.append(arr[middle_index + 1 + i])

#     #comparing and sorting
#     i = j = 0
#     k = left_index

#     while i < left_half_size and j < right_half_size:
#         if temp_left_array[i] <= temp_right_array[j]:
#             arr[k] = temp_left_array[i]
#             i += 1
#         else:
#             arr[k] = temp_right_array[j]
#             j += 1
#         k += 1

#     #copy the remaining elements to the end of array if any
#     while i < left_half_size:
#         arr[k] = temp_left_array[i]
#         i += 1
#         k += 1
#     while j < right_half_size:
#         arr[k] = temp_right_array[j]
#         j += 1
#         k += 1
    

def mergeSort(arr):
    #base case
    if len(arr) == 1:
        return
    
    #create temporary arrays for left and right half
    middle_index = int(math.floor(len(arr) / 2))
    left_half_array = arr[:middle_index]
    right_half_array = arr[middle_index:]
    mergeSort(left_half_array)
    mergeSort(right_half_array)

    #comparing and sorting
    i = j = k = 0
    while i < len(left_half_array) and j < len(right_half_array):
        if left_half_array[i] <= right_half_array[j]:
            arr[k] = left_half_array[i]
            i += 1
        else:
            arr[k] = right_half_array[j]
            j += 1
        k += 1

    #copy the remaining elements to end of array if any
    while i < len(left_half_array):
        arr[k] = left_half_array[i]
        i += 1
        k += 1
    while j < len(right_half_array):
        arr[k] = right_half_array[j]
        j += 1
        k += 1


if __name__ =="__main__":
    main()

File: mergesort.c
#include <cs50.h>
#include <stdio.h>

void mergeSort (int arr[], int left_index, int right_index);

int main (void)
{
    int arr[10] = {46, 27, 96, 1000, 11, -90, 99, 63, 20, 56};

    mergeSort(arr,0,9);

    for (int i = 0; i<10; i++)
    {
        printf("%d, ", arr[i]);
    }

    printf("\n");

}

// O(nlog n) and Omega(nlog n) => theta (nlog n)
// WORKING
void mergeSort (int arr[], int left_index, int right_index)
{
    // return case
    if(left_index == right_index)
    {
        return;
    }

    int middle_index = (left_index+right_index)/2;

    // recursive case
    mergeSort(arr, left_index, middle_index);
    mergeSort(arr, middle_index + 1, right_index);

    // temp arrays for left half and right half
    int n1 = middle_index - left_index + 1;
    int temp_left_arr[n1];
    int n2 = right_index - middle_index;
    int temp_right_arr[n2];
    for (int i = 0; i<n1; i++)
        temp_left_arr[i] = arr[left_index + i];
    for (int j = 0; j<n2; j++)
        temp_right_arr[j] = arr[middle_index + j + 1];

    // comparing and sorting
    int i = 0;
    int j = 0;
    int k = left_index;

    while (i < n1 && j < n2)
    {
        if(temp_left_arr[i] <= temp_right_arr[j])
        {
            arr[k] = temp_left_arr[i];
            i++;
        }
        else
        {
            arr[k] = temp_right_arr[j];
            j++;
        }
        k++;
    }

    // copy remaining elements in temp_left_arr if any
    while (i < n1)
    {
        arr[k] = temp_left_arr[i];
        i++;
        k++;
    }

    // copy remaining elements in temp_right_arr if any
    while (j < n2)
    {
        arr[k] = temp_right_arr[j];
        j++;
        k++;
    }

}